File: ios/Sources/CleverTapNativeDisplay/Models/Background.swift
// MARK: - Background Models
// Background configuration for the Native Display System

import Foundation

/// Background configuration for elements and containers.
/// Supports various background types from simple colors to complex animations.
public enum Background: Codable, Equatable {
    /// Solid color background.
    case solid(SolidBackground)
    
    /// Linear gradient background.
    case linearGradient(LinearGradientBackground)
    
    /// Radial gradient background.
    case radialGradient(RadialGradientBackground)
    
    /// Sweep/Conic gradient background.
    case sweepGradient(SweepGradientBackground)
    
    /// Image background.
    case image(ImageBackground)
    
    /// Shimmer/shine effect background.
    case shimmer(ShimmerBackground)
    
    /// Animated gradient background.
    case animatedGradient(AnimatedGradientBackground)
    
    /// Pulse/breathing effect background.
    case pulse(PulseBackground)
    
    /// Pattern background.
    case pattern(PatternBackground)
    
    /// Particle effect background.
    case particles(ParticlesBackground)
    
    /// Layered background.
    case layered(LayeredBackground)
    
    // MARK: - Codable Implementation
    
    private enum CodingKeys: String, CodingKey {
        case type
    }
    
    public init(from decoder: Decoder) throws {
        let container = try decoder.container(keyedBy: CodingKeys.self)
        let type = try container.decode(String.self, forKey: .type)
        
        switch type {
        case "solid":
            self = .solid(try SolidBackground(from: decoder))
        case "linear_gradient":
            self = .linearGradient(try LinearGradientBackground(from: decoder))
        case "radial_gradient":
            self = .radialGradient(try RadialGradientBackground(from: decoder))
        case "sweep_gradient":
            self = .sweepGradient(try SweepGradientBackground(from: decoder))
        case "image":
            self = .image(try ImageBackground(from: decoder))
        case "shimmer":
            self = .shimmer(try ShimmerBackground(from: decoder))
        case "animated_gradient":
            self = .animatedGradient(try AnimatedGradientBackground(from: decoder))
        case "pulse":
            self = .pulse(try PulseBackground(from: decoder))
        case "pattern":
            self = .pattern(try PatternBackground(from: decoder))
        case "particles":
            self = .particles(try ParticlesBackground(from: decoder))
        case "layered":
            self = .layered(try LayeredBackground(from: decoder))
        default:
            throw DecodingError.dataCorruptedError(
                forKey: .type,
                in: container,
                debugDescription: "Unknown background type: \(type)"
            )
        }
    }
    
    public func encode(to encoder: Encoder) throws {
        switch self {
        case .solid(let bg):
            try bg.encode(to: encoder)
        case .linearGradient(let bg):
            try bg.encode(to: encoder)
        case .radialGradient(let bg):
            try bg.encode(to: encoder)
        case .sweepGradient(let bg):
            try bg.encode(to: encoder)
        case .image(let bg):
            try bg.encode(to: encoder)
        case .shimmer(let bg):
            try bg.encode(to: encoder)
        case .animatedGradient(let bg):
            try bg.encode(to: encoder)
        case .pulse(let bg):
            try bg.encode(to: encoder)
        case .pattern(let bg):
            try bg.encode(to: encoder)
        case .particles(let bg):
            try bg.encode(to: encoder)
        case .layered(let bg):
            try bg.encode(to: encoder)
        }
    }
}

// MARK: - Background Type Structs

public struct SolidBackground: Codable, Equatable {
    public let color: String
    
    public init(color: String) {
        self.color = color
    }
}

public struct LinearGradientBackground: Codable, Equatable {
    public let angle: CGFloat
    public let colors: [String]
    public let stops: [CGFloat]?
    
    public init(angle: CGFloat, colors: [String], stops: [CGFloat]? = nil) {
        self.angle = angle
        self.colors = colors
        self.stops = stops
    }
}

public struct RadialGradientBackground: Codable, Equatable {
    public let centerX: CGFloat
    public let centerY: CGFloat
    public let radius: CGFloat
    public let colors: [String]
    public let stops: [CGFloat]?
    
    private enum CodingKeys: String, CodingKey {
        case centerX = "center_x"
        case centerY = "center_y"
        case radius
        case colors
        case stops
    }
    
    public init(
        centerX: CGFloat = 0.5,
        centerY: CGFloat = 0.5,
        radius: CGFloat = 1.0,
        colors: [String],
        stops: [CGFloat]? = nil
    ) {
        self.centerX = centerX
        self.centerY = centerY
        self.radius = radius
        self.colors = colors
        self.stops = stops
    }

    public init(from decoder: Decoder) throws {
        let c = try decoder.container(keyedBy: CodingKeys.self)
        centerX = try c.decodeIfPresent(CGFloat.self, forKey: .centerX) ?? 0.5
        centerY = try c.decodeIfPresent(CGFloat.self, forKey: .centerY) ?? 0.5
        radius = try c.decodeIfPresent(CGFloat.self, forKey: .radius) ?? 1.0
        colors = try c.decode([String].self, forKey: .colors)
        stops = try c.decodeIfPresent([CGFloat].self, forKey: .stops)
    }
}

public struct SweepGradientBackground: Codable, Equatable {
    public let centerX: CGFloat
    public let centerY: CGFloat
    public let startAngle: CGFloat
    public let colors: [String]
    public let stops: [CGFloat]?
    
    private enum CodingKeys: String, CodingKey {
        case centerX = "center_x"
        case centerY = "center_y"
        case startAngle = "start_angle"
        case colors
        case stops
    }
    
    public init(
        centerX: CGFloat = 0.5,
        centerY: CGFloat = 0.5,
        startAngle: CGFloat = 0,
        colors: [String],
        stops: [CGFloat]? = nil
    ) {
        self.centerX = centerX
        self.centerY = centerY
        self.startAngle = startAngle
        self.colors = colors
        self.stops = stops
    }

    public init(from decoder: Decoder) throws {
        let c = try decoder.container(keyedBy: CodingKeys.self)
        centerX = try c.decodeIfPresent(CGFloat.self, forKey: .centerX) ?? 0.5
        centerY = try c.decodeIfPresent(CGFloat.self, forKey: .centerY) ?? 0.5
        startAngle = try c.decodeIfPresent(CGFloat.self, forKey: .startAngle) ?? 0
        colors = try c.decode([String].self, forKey: .colors)
        stops = try c.decodeIfPresent([CGFloat].self, forKey: .stops)
    }
}

public struct ImageBackground: Codable, Equatable {
    public let url: String
    public let fit: ImageFit
    public let opacity: CGFloat
    public let blur: CGFloat
    public let tint: String?
    public let tintOpacity: CGFloat
    
    private enum CodingKeys: String, CodingKey {
        case url
        case fit
        case opacity
        case blur
        case tint
        case tintOpacity = "tint_opacity"
    }
    
    public init(
        url: String,
        fit: ImageFit = .crop,
        opacity: CGFloat = 1.0,
        blur: CGFloat = 0,
        tint: String? = nil,
        tintOpacity: CGFloat = 0
    ) {
        self.url = url
        self.fit = fit
        self.opacity = opacity
        self.blur = blur
        self.tint = tint
        self.tintOpacity = tintOpacity
    }

    public init(from decoder: Decoder) throws {
        let c = try decoder.container(keyedBy: CodingKeys.self)
        url = try c.decode(String.self, forKey: .url)
        fit = try c.decodeIfPresent(ImageFit.self, forKey: .fit) ?? .crop
        opacity = try c.decodeIfPresent(CGFloat.self, forKey: .opacity) ?? 1.0
        blur = try c.decodeIfPresent(CGFloat.self, forKey: .blur) ?? 0
        tint = try c.decodeIfPresent(String.self, forKey: .tint)
        tintOpacity = try c.decodeIfPresent(CGFloat.self, forKey: .tintOpacity) ?? 0
    }
}

public struct ShimmerBackground: Codable, Equatable {
    public let baseColor: String
    public let highlightColor: String
    public let angle: CGFloat
    public let duration: Int
    public let loop: Bool
    
    private enum CodingKeys: String, CodingKey {
        case baseColor = "base_color"
        case highlightColor = "highlight_color"
        case angle
        case duration
        case loop
    }
    
    public init(
        baseColor: String,
        highlightColor: String,
        angle: CGFloat = 45,
        duration: Int = 1500,
        loop: Bool = true
    ) {
        self.baseColor = baseColor
        self.highlightColor = highlightColor
        self.angle = angle
        self.duration = duration
        self.loop = loop
    }

    public init(from decoder: Decoder) throws {
        let c = try decoder.container(keyedBy: CodingKeys.self)
        baseColor = try c.decode(String.self, forKey: .baseColor)
        highlightColor = try c.decode(String.self, forKey: .highlightColor)
        angle = try c.decodeIfPresent(CGFloat.self, forKey: .angle) ?? 45
        duration = try c.decodeIfPresent(Int.self, forKey: .duration) ?? 1500
        loop = try c.decodeIfPresent(Bool.self, forKey: .loop) ?? true
    }
}

public struct AnimatedGradientBackground: Codable, Equatable {
    public let gradientType: GradientType
    public let angle: CGFloat
    public let colors: [String]
    public let duration: Int
    public let loop: Bool
    public let animationStyle: AnimationStyle
    
    private enum CodingKeys: String, CodingKey {
        case gradientType = "gradient_type"
        case angle
        case colors
        case duration
        case loop
        case animationStyle = "animation_style"
    }
    
    public init(
        gradientType: GradientType,
        angle: CGFloat = 0,
        colors: [String],
        duration: Int = 3000,
        loop: Bool = true,
        animationStyle: AnimationStyle = .smooth
    ) {
        self.gradientType = gradientType
        self.angle = angle
        self.colors = colors
        self.duration = duration
        self.loop = loop
        self.animationStyle = animationStyle
    }

    public init(from decoder: Decoder) throws {
        let c = try decoder.container(keyedBy: CodingKeys.self)
        gradientType = try c.decode(GradientType.self, forKey: .gradientType)
        angle = try c.decodeIfPresent(CGFloat.self, forKey: .angle) ?? 0
        colors = try c.decode([String].self, forKey: .colors)
        duration = try c.decodeIfPresent(Int.self, forKey: .duration) ?? 3000
        loop = try c.decodeIfPresent(Bool.self, forKey: .loop) ?? true
        animationStyle = try c.decodeIfPresent(AnimationStyle.self, forKey: .animationStyle) ?? .smooth
    }
}

public struct PulseBackground: Codable, Equatable {
    public let color: String
    public let minOpacity: CGFloat
    public let maxOpacity: CGFloat
    public let duration: Int
    public let loop: Bool
    
    private enum CodingKeys: String, CodingKey {
        case color
        case minOpacity = "min_opacity"
        case maxOpacity = "max_opacity"
        case duration
        case loop
    }
    
    public init(
        color: String,
        minOpacity: CGFloat = 0.3,
        maxOpacity: CGFloat = 1.0,
        duration: Int = 1000,
        loop: Bool = true
    ) {
        self.color = color
        self.minOpacity = minOpacity
        self.maxOpacity = maxOpacity
        self.duration = duration
        self.loop = loop
    }

    public init(from decoder: Decoder) throws {
        let c = try decoder.container(keyedBy: CodingKeys.self)
        color = try c.decode(String.self, forKey: .color)
        minOpacity = try c.decodeIfPresent(CGFloat.self, forKey: .minOpacity) ?? 0.3
        maxOpacity = try c.decodeIfPresent(CGFloat.self, forKey: .maxOpacity) ?? 1.0
        duration = try c.decodeIfPresent(Int.self, forKey: .duration) ?? 1000
        loop = try c.decodeIfPresent(Bool.self, forKey: .loop) ?? true
    }
}

public struct PatternBackground: Codable, Equatable {
    public let patternType: PatternType
    public let primaryColor: String
    public let secondaryColor: String
    public let size: CGFloat
    public let spacing: CGFloat
    public let opacity: CGFloat
    
    private enum CodingKeys: String, CodingKey {
        case patternType = "pattern_type"
        case primaryColor = "primary_color"
        case secondaryColor = "secondary_color"
        case size
        case spacing
        case opacity
    }
    
    public init(
        patternType: PatternType,
        primaryColor: String,
        secondaryColor: String,
        size: CGFloat = 20,
        spacing: CGFloat = 30,
        opacity: CGFloat = 1.0
    ) {
        self.patternType = patternType
        self.primaryColor = primaryColor
        self.secondaryColor = secondaryColor
        self.size = size
        self.spacing = spacing
        self.opacity = opacity
    }

    public init(from decoder: Decoder) throws {
        let c = try decoder.container(keyedBy: CodingKeys.self)
        patternType = try c.decode(PatternType.self, forKey: .patternType)
        primaryColor = try c.decode(String.self, forKey: .primaryColor)
        secondaryColor = try c.decode(String.self, forKey: .secondaryColor)
        size = try c.decodeIfPresent(CGFloat.self, forKey: .size) ?? 20
        spacing = try c.decodeIfPresent(CGFloat.self, forKey: .spacing) ?? 30
        opacity = try c.decodeIfPresent(CGFloat.self, forKey: .opacity) ?? 1.0
    }
}

public struct ParticlesBackground: Codable, Equatable {
    public let particleColor: String
    public let particleCount: Int
    public let particleSize: CGFloat
    public let speed: CGFloat
    public let direction: ParticleDirection
    public let opacity: CGFloat
    
    private enum CodingKeys: String, CodingKey {
        case particleColor = "particle_color"
        case particleCount = "particle_count"
        case particleSize = "particle_size"
        case speed
        case direction
        case opacity
    }
    
    public init(
        particleColor: String,
        particleCount: Int = 50,
        particleSize: CGFloat = 4,
        speed: CGFloat = 2,
        direction: ParticleDirection = .up,
        opacity: CGFloat = 0.7
    ) {
        self.particleColor = particleColor
        self.particleCount = particleCount
        self.particleSize = particleSize
        self.speed = speed
        self.direction = direction
        self.opacity = opacity
    }

    public init(from decoder: Decoder) throws {
        let c = try decoder.container(keyedBy: CodingKeys.self)
        particleColor = try c.decode(String.self, forKey: .particleColor)
        particleCount = try c.decodeIfPresent(Int.self, forKey: .particleCount) ?? 50
        particleSize = try c.decodeIfPresent(CGFloat.self, forKey: .particleSize) ?? 4
        speed = try c.decodeIfPresent(CGFloat.self, forKey: .speed) ?? 2
        direction = try c.decodeIfPresent(ParticleDirection.self, forKey: .direction) ?? .up
        opacity = try c.decodeIfPresent(CGFloat.self, forKey: .opacity) ?? 0.7
    }
}

public struct LayeredBackground: Codable, Equatable {
    public let layers: [Background]
    
    public init(layers: [Background]) {
        self.layers = layers
    }
}

File: flutter/lib/src/models/background.dart
import 'enums.dart';

sealed class Background {
  const Background();

  factory Background.fromJson(Map<String, dynamic> json) {
    final type = json['type'] as String? ?? '';
    return switch (type) {
      'solid' => SolidBackground.fromJson(json),
      'linear_gradient' => LinearGradientBackground.fromJson(json),
      'radial_gradient' => RadialGradientBackground.fromJson(json),
      'sweep_gradient' => SweepGradientBackground.fromJson(json),
      'image' => ImageBackground.fromJson(json),
      'shimmer' => ShimmerBackground.fromJson(json),
      'animated_gradient' => AnimatedGradientBackground.fromJson(json),
      'pulse' => PulseBackground.fromJson(json),
      'pattern' => PatternBackground.fromJson(json),
      'particles' => ParticlesBackground.fromJson(json),
      'layered' => LayeredBackground.fromJson(json),
      _ => SolidBackground(color: '#00000000'),
    };
  }
}

class SolidBackground extends Background {
  final String color;

  const SolidBackground({required this.color});

  factory SolidBackground.fromJson(Map<String, dynamic> json) =>
      SolidBackground(color: json['color'] as String? ?? '#000000');
}

class LinearGradientBackground extends Background {
  final double angle;
  final List<String> colors;
  final List<double>? stops;

  const LinearGradientBackground({
    required this.angle,
    required this.colors,
    this.stops,
  });

  factory LinearGradientBackground.fromJson(Map<String, dynamic> json) =>
      LinearGradientBackground(
        angle: (json['angle'] as num?)?.toDouble() ?? 0,
        colors: (json['colors'] as List?)?.map((e) => e as String).toList() ?? const [],
        stops: (json['stops'] as List?)?.map((e) => (e as num).toDouble()).toList(),
      );
}

class RadialGradientBackground extends Background {
  final double centerX;
  final double centerY;
  final double radius;
  final List<String> colors;
  final List<double>? stops;

  const RadialGradientBackground({
    this.centerX = 0.5,
    this.centerY = 0.5,
    this.radius = 1.0,
    required this.colors,
    this.stops,
  });

  factory RadialGradientBackground.fromJson(Map<String, dynamic> json) =>
      RadialGradientBackground(
        centerX: (json['center_x'] as num?)?.toDouble() ?? 0.5,
        centerY: (json['center_y'] as num?)?.toDouble() ?? 0.5,
        radius: (json['radius'] as num?)?.toDouble() ?? 1.0,
        colors: (json['colors'] as List?)?.map((e) => e as String).toList() ?? const [],
        stops: (json['stops'] as List?)?.map((e) => (e as num).toDouble()).toList(),
      );
}

class SweepGradientBackground extends Background {
  final double centerX;
  final double centerY;
  final double startAngle;
  final List<String> colors;
  final List<double>? stops;

  const SweepGradientBackground({
    this.centerX = 0.5,
    this.centerY = 0.5,
    this.startAngle = 0,
    required this.colors,
    this.stops,
  });

  factory SweepGradientBackground.fromJson(Map<String, dynamic> json) =>
      SweepGradientBackground(
        centerX: (json['center_x'] as num?)?.toDouble() ?? 0.5,
        centerY: (json['center_y'] as num?)?.toDouble() ?? 0.5,
        startAngle: (json['start_angle'] as num?)?.toDouble() ?? 0,
        colors: (json['colors'] as List?)?.map((e) => e as String).toList() ?? const [],
        stops: (json['stops'] as List?)?.map((e) => (e as num).toDouble()).toList(),
      );
}

class ImageBackground extends Background {
  final String url;
  final ImageFit fit;
  final double opacity;
  final double blur;
  final String? tint;
  final double tintOpacity;

  const ImageBackground({
    required this.url,
    this.fit = ImageFit.crop,
    this.opacity = 1.0,
    this.blur = 0,
    this.tint,
    this.tintOpacity = 0,
  });

  factory ImageBackground.fromJson(Map<String, dynamic> json) => ImageBackground(
        url: json['url'] as String? ?? '',
        fit: ImageFit.fromJson(json['fit'] as String? ?? 'crop'),
        opacity: (json['opacity'] as num?)?.toDouble() ?? 1.0,
        blur: (json['blur'] as num?)?.toDouble() ?? 0,
        tint: json['tint'] as String?,
        tintOpacity: (json['tint_opacity'] as num?)?.toDouble() ?? 0,
      );
}

// Deferred to v2 — parsed but rendered as transparent with warning
class ShimmerBackground extends Background {
  final String baseColor;
  final String highlightColor;
  final double angle;
  final int duration;
  final bool loop;

  const ShimmerBackground({
    required this.baseColor,
    required this.highlightColor,
    this.angle = 45,
    this.duration = 1500,
    this.loop = true,
  });

  factory ShimmerBackground.fromJson(Map<String, dynamic> json) => ShimmerBackground(
        baseColor: json['base_color'] as String? ?? '#E0E0E0',
        highlightColor: json['highlight_color'] as String? ?? '#F5F5F5',
        angle: (json['angle'] as num?)?.toDouble() ?? 45,
        duration: json['duration'] as int? ?? 1500,
        loop: json['loop'] as bool? ?? true,
      );
}

// Deferred to v2 — parsed but rendered as transparent with warning
class AnimatedGradientBackground extends Background {
  final GradientType gradientType;
  final double angle;
  final List<String> colors;
  final int duration;
  final bool loop;
  final AnimationStyle animationStyle;

  const AnimatedGradientBackground({
    required this.gradientType,
    this.angle = 0,
    required this.colors,
    this.duration = 3000,
    this.loop = true,
    this.animationStyle = AnimationStyle.smooth,
  });

  factory AnimatedGradientBackground.fromJson(Map<String, dynamic> json) =>
      AnimatedGradientBackground(
        gradientType: GradientType.fromJson(json['gradient_type'] as String? ?? 'linear'),
        angle: (json['angle'] as num?)?.toDouble() ?? 0,
        colors: (json['colors'] as List?)?.map((e) => e as String).toList() ?? const [],
        duration: json['duration'] as int? ?? 3000,
        loop: json['loop'] as bool? ?? true,
        animationStyle: AnimationStyle.fromJson(json['animation_style'] as String? ?? 'smooth'),
      );
}

// Deferred to v2 — parsed but rendered as transparent with warning
class PulseBackground extends Background {
  final String color;
  final double minOpacity;
  final double maxOpacity;
  final int duration;
  final bool loop;

  const PulseBackground({
    required this.color,
    this.minOpacity = 0.3,
    this.maxOpacity = 1.0,
    this.duration = 1000,
    this.loop = true,
  });

  factory PulseBackground.fromJson(Map<String, dynamic> json) => PulseBackground(
        color: json['color'] as String? ?? '#000000',
        minOpacity: (json['min_opacity'] as num?)?.toDouble() ?? 0.3,
        maxOpacity: (json['max_opacity'] as num?)?.toDouble() ?? 1.0,
        duration: json['duration'] as int? ?? 1000,
        loop: json['loop'] as bool? ?? true,
      );
}

// Deferred to v2 — parsed but rendered as transparent with warning
class PatternBackground extends Background {
  final PatternType patternType;
  final String primaryColor;
  final String secondaryColor;
  final double size;
  final double spacing;
  final double opacity;

  const PatternBackground({
    required this.patternType,
    required this.primaryColor,
    required this.secondaryColor,
    this.size = 20,
    this.spacing = 30,
    this.opacity = 1.0,
  });

  factory PatternBackground.fromJson(Map<String, dynamic> json) => PatternBackground(
        patternType: PatternType.fromJson(json['pattern_type'] as String? ?? 'dots'),
        primaryColor: json['primary_color'] as String? ?? '#000000',
        secondaryColor: json['secondary_color'] as String? ?? '#FFFFFF',
        size: (json['size'] as num?)?.toDouble() ?? 20,
        spacing: (json['spacing'] as num?)?.toDouble() ?? 30,
        opacity: (json['opacity'] as num?)?.toDouble() ?? 1.0,
      );
}

// Deferred to v2 — parsed but rendered as transparent with warning
class ParticlesBackground extends Background {
  final String particleColor;
  final int particleCount;
  final double particleSize;
  final double speed;
  final ParticleDirection direction;
  final double opacity;

  const ParticlesBackground({
    required this.particleColor,
    this.particleCount = 50,
    this.particleSize = 4,
    this.speed = 2,
    this.direction = ParticleDirection.up,
    this.opacity = 0.7,
  });

  factory ParticlesBackground.fromJson(Map<String, dynamic> json) => ParticlesBackground(
        particleColor: json['particle_color'] as String? ?? '#000000',
        particleCount: json['particle_count'] as int? ?? 50,
        particleSize: (json['particle_size'] as num?)?.toDouble() ?? 4,
        speed: (json['speed'] as num?)?.toDouble() ?? 2,
        direction: ParticleDirection.fromJson(json['direction'] as String? ?? 'up'),
        opacity: (json['opacity'] as num?)?.toDouble() ?? 0.7,
      );
}

// Deferred to v2 — parsed but rendered as transparent with warning
class LayeredBackground extends Background {
  final List<Background> layers;

  const LayeredBackground({required this.layers});

  factory LayeredBackground.fromJson(Map<String, dynamic> json) => LayeredBackground(
        layers: (json['layers'] as List?)
                ?.map((e) => Background.fromJson(e as Map<String, dynamic>))
                .toList() ??
            const [],
      );
}
